src/NServiceBus.Core/Pipeline/Incoming/IUnitOfWorkContext.cs

```diff
@@ -0,0 +1,38 @@
+namespace NServiceBus.Pipeline
+{
+    using System.Collections.Generic;
+    using Persistence;
+    using Unicast.Messages;
+
+    /// <summary>
+    /// A context for creating setting up custom unit of work for handlers.
+    /// </summary>
+    public interface IUnitOfWorkContext : IIncomingContext
+    {
+        /// <summary>
+        /// Gets the synchronized storage session for processing the current message. NServiceBus makes sure the changes made
+        /// via this session will be persisted before the message receive is acknowledged.
+        /// </summary>
+        SynchronizedStorageSession SynchronizedStorageSession { get; }
+
+        /// <summary>
+        /// The message instance being handled.
+        /// </summary>
+        object MessageBeingHandled { get; }
+
+        /// <summary>
+        /// Message headers.
+        /// </summary>
+        Dictionary<string, string> Headers { get; }
+
+        /// <summary>
+        /// Metadata for the incoming message.
+        /// </summary>
+        MessageMetadata MessageMetadata { get; }
+
+        /// <summary>
+        /// Tells if the message has been handled.
+        /// </summary>
+        bool MessageHandled { get; set; }
```

Comment:
Why do we need this here? 

Pretty sure it was added as a hack to support the callbacks flagging a message as handled

https://github.com/Particular/NServiceBus.Callbacks/blob/7d69f94979e7dc82c8527d7cbc0a0d95a863a7e0/src/NServiceBus.Callbacks/Reply/RequestResponseInvocationForMessagesBehavior.cs#L31

to avoid us throwing a "message wasn't handled" exception.

I don't think we need that hack here?